core/src/main/java/com/microsoft/applicationinsights/internal/statsbeat/AttachStatsbeat.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * ApplicationInsights-Java
+ * Copyright (c) Microsoft Corporation
+ * All rights reserved.
+ *
+ * MIT License
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this
+ * software and associated documentation files (the ""Software""), to deal in the Software
+ * without restriction, including without limitation the rights to use, copy, modify, merge,
+ * publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+ * persons to whom the Software is furnished to do so, subject to the following conditions:
+ * The above copyright notice and this permission notice shall be included in all copies or
+ * substantial portions of the Software.
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+ * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+ * FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+package com.microsoft.applicationinsights.internal.statsbeat;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.microsoft.applicationinsights.TelemetryClient;
+import com.microsoft.applicationinsights.telemetry.MetricTelemetry;
+
+import static com.microsoft.applicationinsights.internal.statsbeat.Constants.*;
+
+public class AttachStatsbeat extends BaseStatsbeat {
+
+    private String resourceProviderId;
+    private MetadataInstanceResponse metadataInstanceResponse;
+
+    public AttachStatsbeat(TelemetryClient telemetryClient, long interval) {
+        super(telemetryClient, interval);
+        initResourceProviderId();
+        AzureMetadataService.getInstance().initialize(interval);
+    }
+
+    @Override
+    protected void send() {
+        if (resourceProviderId != null) {
+            MetricTelemetry statsbeatTelemetry = createStatsbeatTelemetry(ATTACH, 0);
+            statsbeatTelemetry.getProperties().put(CUSTOM_DIMENSIONS_RP_ID, resourceProviderId);
+            telemetryClient.track(statsbeatTelemetry);
+        }
+    }
+
+    /**
+     * @return the unique identifier of the resource provider.
+     */
+    public String getResourceProviderId() {
+        return resourceProviderId;
+    }
+
+    public MetadataInstanceResponse getMetadataInstanceResponse() {
+        return metadataInstanceResponse;
+    }
+
+    public void updateMetadataInstance(MetadataInstanceResponse response) {
+        if (response != null) {
+            metadataInstanceResponse = response;
+            updateResourceProvider(RP_VM);
+            updateOperatingSystem();
+        }
+    }
+
+    protected void updateResourceProvider(String rp) {
+        CustomDimensions.getInstance().getProperties().put(CUSTOM_DIMENSIONS_RP, rp);
```

Comment:
I think it's a little confusing to have CustomDimensions updated from inside of AttachStatsbeat, since it also needs to be updated for the others, I think would be clearer to treat CustomDimensions as its own top-level concept and set it directly inside of parseJsonResponse